driver/src/orderbook/streamed/state.rs

```diff
@@ -79,57 +81,47 @@ impl State {
 
     /// Apply an event to the state, modifying it.
     ///
-    /// In case of error no modifications take place.
+    /// Consumes and returns back `self` because it cannot be reused in case of error.
     ///
     /// `block_batch_id` is the current batch based on the timestamp of the block that contains the
     ///  event.
-    pub fn apply_event(&mut self, event: &Event, block_batch_id: BatchId) -> Result<(), Error> {
+    pub fn apply_event(mut self, event: &Event, block_batch_id: BatchId) -> Result<Self> {
```

Comment:
This very cool! Passing self by value ensures that we don't use invalid states on errors. I love it!